Mathlib/LinearAlgebra/TensorProduct/Finiteness.lean

```diff
@@ -0,0 +1,301 @@
+  `TensorProduct.exists_finite_submodule_of_fintype` and 21 more:
+  any finitely many elements `M ⊗[R] N` is contained in `M' ⊗[R] N'`
+  for some finitely generated submodules `M'` and `N'` of `M` and `N`, respectively.
+  Each of these 3 functions has 8 variants.
+
+## Tags
+
+tensor product, finitely generated
+
+-/
+
+open scoped Classical TensorProduct
+
+open Submodule
+
+noncomputable section
+
+variable {R M N : Type*}
+
+variable [CommSemiring R] [AddCommMonoid M] [AddCommMonoid N] [Module R M] [Module R N]
+
+variable {M₁ : Submodule R M} {N₁ : Submodule R N}
+
+namespace TensorProduct
+
+/-- For any element `x` of `M ⊗[R] N`, there exists a (finite) multiset `{ (m_i, n_i) }`
+of `M × N`, such that `x` is equal to the sum of `m_i ⊗ₜ[R] n_i`. -/
+theorem exists_multiset (x : M ⊗[R] N) :
+    ∃ S : Multiset (M × N), x = (S.map fun i ↦ i.1 ⊗ₜ[R] i.2).sum := by
+  induction x using TensorProduct.induction_on with
+  | zero => exact ⟨0, by simp⟩
+  | tmul x y => exact ⟨{(x, y)}, by simp⟩
+  | add x y hx hy =>
+    obtain ⟨Sx, hx⟩ := hx
+    obtain ⟨Sy, hy⟩ := hy
+    exact ⟨Sx + Sy, by rw [Multiset.map_add, Multiset.sum_add, hx, hy]⟩
+
+/-- For any element `x` of `M ⊗[R] N`, there exists a finite subset `{ (m_i, n_i) }`
+of `M × N` such that each `m_i` is distinct (we represent it as an element of `M →₀ N`),
+such that `x` is equal to the sum of `m_i ⊗ₜ[R] n_i`. -/
+theorem exists_finsupp_left (x : M ⊗[R] N) :
+    ∃ S : M →₀ N, x = S.sum fun m n ↦ m ⊗ₜ[R] n := by
+  induction x using TensorProduct.induction_on with
+  | zero => exact ⟨0, by simp⟩
+  | tmul x y => exact ⟨Finsupp.single x y, by simp⟩
+  | add x y hx hy =>
+    obtain ⟨Sx, hx⟩ := hx
+    obtain ⟨Sy, hy⟩ := hy
+    use Sx + Sy
+    rw [hx, hy]
+    exact (Finsupp.sum_add_index' (by simp) TensorProduct.tmul_add).symm
+
+/-- For any element `x` of `M ⊗[R] N`, there exists a finite subset `{ (m_i, n_i) }`
+of `M × N` such that each `n_i` is distinct (we represent it as an element of `N →₀ M`),
+such that `x` is equal to the sum of `m_i ⊗ₜ[R] n_i`. -/
+theorem exists_finsupp_right (x : M ⊗[R] N) :
+    ∃ S : N →₀ M, x = S.sum fun n m ↦ m ⊗ₜ[R] n := by
+  obtain ⟨S, h⟩ := exists_finsupp_left (TensorProduct.comm R M N x)
+  refine ⟨S, (TensorProduct.comm R M N).injective ?_⟩
+  simp_rw [h, Finsupp.sum, map_sum]; rfl
+
+/-- For any element `x` of `M ⊗[R] N`, there exists a finite subset `{ (m_i, n_i) }`
+of `M × N`, such that `x` is equal to the sum of `m_i ⊗ₜ[R] n_i`. -/
+theorem exists_finset (x : M ⊗[R] N) :
+    ∃ S : Finset (M × N), x = S.sum fun i ↦ i.1 ⊗ₜ[R] i.2 := by
+  obtain ⟨S, h⟩ := exists_finsupp_left x
+  use S.graph
+  rw [h, Finsupp.sum]
+  refine' Finset.sum_nbij' (fun m ↦ ⟨m, S m⟩) Prod.fst .. <;> simp
+
+/-- If `x : ι → M ⊗[R] N` are finitely many elements, then there exists a finitely generated
+submodule `M'` of `M` and `x' : ι → M' ⊗[R] N`, whose image in `M ⊗[R] N` is equal to `x`.
+There are 8 variants of this function. -/
```

Comment:
Do we really want 8 versions of this result? I understand that finding a single canonical one is probably not possible, but 8 seems really a lot.

Reply:
I think 
```
example (s : Set (M ⊗[R] N)) (hs : s.Finite) : 
    ∃ (M' : Submodule R M) (N' : Submodule R N), M'.FG ∧ N'.FG ∧
      s ⊆ (TensorProduct.map M'.subtype N'.subtype).range
```
is enough?

Reply:
Also I don't think the lemmas above are necessary. Usually when you find yourself writing tensors as a sum of pure tensors in lean, it is usually better done directly using induction.
```lean
import Mathlib

variable {R M N : Type*}

variable [CommSemiring R] [AddCommMonoid M] [AddCommMonoid N] [Module R M] [Module R N]

variable {M₁ : Submodule R M} {N₁ : Submodule R N}

open TensorProduct Submodule

lemma TensorProduct.range_mapIncl (M' : Submodule R M) (N' : Submodule R N) :
    LinearMap.range (mapIncl M' N') = span R (Set.image2 (· ⊗ₜ ·) M' N') := by
  rw [mapIncl, map_range_eq_span_tmul]
  congr; ext; simp

lemma TensorProduct.range_mapIncl_mono
    {M₁' M₂' : Submodule R M} {N₁' N₂' : Submodule R N}
    (hM : M₁' ≤ M₂') (hN : N₁' ≤ N₂') :
    LinearMap.range (mapIncl M₁' N₁') ≤ LinearMap.range (mapIncl M₂' N₂') := by
  rw [range_mapIncl, range_mapIncl]
  exact span_mono (Set.image2_subset (SetLike.le_def.mp hM) (SetLike.le_def.mp hN))

example (s : Set (M ⊗[R] N)) (hs : s.Finite) : 
    ∃ (M' : Submodule R M) (N' : Submodule R N), M'.FG ∧ N'.FG ∧
      s ⊆ LinearMap.range (mapIncl M' N') := by
  refine hs.induction_on ?_ ?_
  · exact ⟨_, _, fg_bot, fg_bot, Set.empty_subset _⟩
  rintro a s - - ⟨M', N', hM', hN', h⟩
  induction' a using TensorProduct.induction_on with x y x y hx hy
  · exact ⟨M', N', hM', hN', Set.insert_subset (zero_mem _) h⟩
  · refine ⟨_, _, hM'.sup (fg_span_singleton x),
      hN'.sup (fg_span_singleton y), Set.insert_subset ?_ ?_⟩
    · exact ⟨⟨x, mem_sup_right (mem_span_singleton_self x)⟩ ⊗ₜ
        ⟨y, mem_sup_right (mem_span_singleton_self y)⟩, rfl⟩
    · exact fun z hz ↦ range_mapIncl_mono le_sup_left le_sup_left (h hz)
  · obtain ⟨M₁', N₁', hM₁', hN₁', h₁⟩ := hx
    obtain ⟨M₂', N₂', hM₂', hN₂', h₂⟩ := hy
    refine ⟨_, _, hM₁'.sup hM₂', hN₁'.sup hN₂', Set.insert_subset (add_mem ?_ ?_) ?_⟩
    · exact range_mapIncl_mono le_sup_left le_sup_left (h₁ (Set.mem_insert x s))
    · exact range_mapIncl_mono le_sup_right le_sup_right (h₂ (Set.mem_insert y s))
    · exact fun z hz ↦ range_mapIncl_mono le_sup_left le_sup_left
        (h₁ (Set.subset_insert x s hz))
```

Reply:
> Do we really want 8 versions of this result? I understand that finding a single canonical one is probably not possible, but 8 seems really a lot.

Let me see. Currently there are the following combinations:

- submodules: left, right or both
- M and N are types with module structure, or they are already submodules
- fintype, finsupp, pair, or only one element. From the proof I think the latter two cases can be removed. Also the finsupp case is a possible variation, but it's never used yet.

In fact what I really need is `exists_finite_submodule_left_of_pair'` and `exists_finite_submodule_right_of_pair'`, used in https://github.com/leanprover-community/mathlib4/blob/2d07f5a79f80dd8d4371b2bc02267ec03b3c88c5/Mathlib/LinearAlgebra/LinearDisjoint.lean#L483-L502

Reply:
OK, now there are only 6 results of `exists_finite_submodule`. Adapted the proof of @erdOne. Please give some comments.

> And I don't think there is need for left and right versions.

Unfortunately I need to use them in linear disjoint file... However now left and right versions are corollaries of the both version.

Also the `exists_multiset`, etc. are not removed yet. I think they *may be* useful by someone...

Reply:
I originally thought that the `LinearMap.range_comp_le_range` part could be performed inline at the application site and there is no need for separate lemmas concerning left and right. But as long as the number of lemmas does not overwhelm the user, we might as well provide them for convenience.